PocketKit/Sources/SharedPocketKit/Session/SessionBackupUtility.swift

```diff
@@ -0,0 +1,113 @@
+
+    private let userDefaults: UserDefaults
+    private let store: LegacyEncryptedStore
+    private let notificationCenter: NotificationCenter
+
+    private let notificationQueue = DispatchQueue(label: "SessionBackupUtility", qos: .utility)
+    private var subscriptions: Set<AnyCancellable> = []
+
+    public init(userDefaults: UserDefaults, store: LegacyEncryptedStore, notificationCenter: NotificationCenter) {
+        self.userDefaults = userDefaults
+        self.store = store
+        self.notificationCenter = notificationCenter
+
+        // Return an existing decryption key, else create + store a new one, and return that
+        if let existing = userDefaults.string(forKey: LegacyUserMigration.decryptionKey) {
+            self.password = existing
+        } else {
+            // Generate a password the same way as the legacy app
+            let password = RNCryptor.randomData(ofLength: 32).base64EncodedString()
+            userDefaults.set(password, forKey: LegacyUserMigration.decryptionKey)
+            self.password = password
+        }
+    }
+
+    public func start() {
+        // Utilize NotificationCenter to save a user session on-disk after a user has logged in
+        notificationCenter.publisher(for: .userLoggedIn)
+            .receive(on: notificationQueue)
+            .sink { [weak self] notification in
+                guard let session = notification.object as? Session else {
+                    return
+                }
+
+                try? self?.save(session: session)
+            }.store(in: &subscriptions)
+
+        // Utilize NotificationCenter to reset a user session on-disk after a user has logged out
+        notificationCenter.publisher(for: .userLoggedOut)
+            .receive(on: notificationQueue)
+            .sink { [weak self] notification in
+                try? self?.save(session: nil)
+            }.store(in: &subscriptions)
+    }
+
+    /// Saves a `SharedPocketKit.Session` to `store`, adding / modifying existing key/value pairs
+    ///  within the store to contain the requested Session data.
+    private func save(session: Session?) throws {
+        let data: Data
+        do {
+            // `decryptStore` will throw if the on-disk store was
+            // empty, or if something went wrong. We want to continue saving
+            // the session, though, so we will default to empty data and continue.
+            // This assumes that if the file already existed (via Pocket 7), that this won't throw,
+            // and will only throw if the file was empty (i.e empty when created by Pocket 8).
+            data = try store.decryptStore(securedBy: password) ?? Data()
+        } catch {
+            data = Data()
+        }
+
+        let initial: [String: Any]
+
+        if data.isEmpty { // data will be empty if the store has been created in Pocket 8
+            initial = [:]
+        } else { // data will not be empty if previously created, either in Pocket 7 or 8
```

Comment:
Should we add a version field or similar to an entry in the file to be explicit and use that?

Reply:
I'm not sure it necessarily matters _which_ version it was created in, within the context of this utility. The file either is there, or isn't, which is what's more crucial. However, we _can_ update the version stored to the version of the application, but not use it.